sycl/unittests/kernel-and-program/Cache.cpp

```diff
@@ -0,0 +1,485 @@
+  Mock.redefine<detail::PiApiKind::piKernelRelease>(redefinedKernelRelease);
+  Mock.redefine<detail::PiApiKind::piKernelGetInfo>(redefinedKernelGetInfo);
+
+  context Ctx{Plt};
+  program Prg{Ctx};
+
+  Prg.build_with_source("", "-g");
+  auto CtxImpl = detail::getSyclObjImpl(Ctx);
+  detail::KernelProgramCache::ProgramCacheT &Cache =
+      CtxImpl->getKernelProgramCache().acquireCachedPrograms().get();
+  EXPECT_EQ(Cache.size(), 0) << "Expect empty cache for source programs";
+}
+
+TEST(KernelAndProgramCache, ProgramSourceNegativeCompileAndLink) {
+  platform Plt{default_selector()};
+  if (Plt.is_host() || Plt.get_backend() != backend::opencl) {
+    std::clog << "This test is only supported on OpenCL devices\n";
+    return;
+  }
+
+  unittest::PiMock Mock(Plt);
+  Mock.redefine<detail::PiApiKind::piclProgramCreateWithSource>(
+      redefinedProgramCreateWithSource);
+  Mock.redefine<detail::PiApiKind::piProgramCompile>(redefinedProgramCompile);
+  Mock.redefine<detail::PiApiKind::piProgramLink>(redefinedProgramLink);
+  Mock.redefine<detail::PiApiKind::piKernelCreate>(redefinedKernelCreate);
+  Mock.redefine<detail::PiApiKind::piKernelRetain>(redefinedKernelRetain);
+  Mock.redefine<detail::PiApiKind::piKernelRelease>(redefinedKernelRelease);
+  Mock.redefine<detail::PiApiKind::piKernelGetInfo>(redefinedKernelGetInfo);
+
+  context Ctx{Plt};
+  program Prg{Ctx};
+
+  Prg.compile_with_source("");
+  Prg.link();
+  auto CtxImpl = detail::getSyclObjImpl(Ctx);
+  detail::KernelProgramCache::ProgramCacheT &Cache =
+      CtxImpl->getKernelProgramCache().acquireCachedPrograms().get();
+  EXPECT_EQ(Cache.size(), 0) << "Expect empty cache for source programs";
+}
+
+TEST(KernelAndProgramCache, ProgramSourceNegativeCompileAndLinkWithOpts) {
+  platform Plt{default_selector()};
+  if (Plt.is_host() || Plt.get_backend() != backend::opencl) {
+    std::clog << "This test is only supported on OpenCL devices\n";
+    return;
+  }
+
+  unittest::PiMock Mock(Plt);
+  Mock.redefine<detail::PiApiKind::piclProgramCreateWithSource>(
+      redefinedProgramCreateWithSource);
+  Mock.redefine<detail::PiApiKind::piProgramCompile>(redefinedProgramCompile);
+  Mock.redefine<detail::PiApiKind::piProgramLink>(redefinedProgramLink);
+  Mock.redefine<detail::PiApiKind::piKernelCreate>(redefinedKernelCreate);
+  Mock.redefine<detail::PiApiKind::piKernelRetain>(redefinedKernelRetain);
+  Mock.redefine<detail::PiApiKind::piKernelRelease>(redefinedKernelRelease);
+  Mock.redefine<detail::PiApiKind::piKernelGetInfo>(redefinedKernelGetInfo);
+
+  context Ctx{Plt};
+  program Prg{Ctx};
+
+  Prg.compile_with_source("");
+  Prg.link();
+  auto CtxImpl = detail::getSyclObjImpl(Ctx);
+  detail::KernelProgramCache::ProgramCacheT &Cache =
+      CtxImpl->getKernelProgramCache().acquireCachedPrograms().get();
+  EXPECT_EQ(Cache.size(), 0) << "Expect empty cache for source programs";
+}
+
+TEST(KernelAndProgramCache, ProgramBuildPositive) {
```

Comment:
Kernel-cache tests start with `KernelCache` while program-cache tests start with `Program` only.
Please align names of the tests for kernel cache with program cache.